### DISCUSSION TITLE
8장

### DISCUSSION THREAD

**Discussion by rawfishthelgh** [somefood/cs-study#67] [General]
<h2>테이블 설계의 기초</h2>
<h3>RDBMS가 주류가 된 이유</h3>
<p>rdbms가 표준 데이터베이스 제품이 된 이유는 데이터 정합성을 위한 “설계 노하우”가 매우 발달했기 때문이다. 즉 데이터베이스에서 정합성을 매우 높은 단계로 요구되고, 정합성을 높이기 위한 테이블 설계는 매우 중요함.</p>
<h3>테이블 설계는 논리의 세계</h3>
<p>시스템에서 하드웨어의 세계를 “물리”, 소프트웨어의 설계를 “논리”로 일컫는다. 테이블 설계는 서버나 스토리지 같은 물리층과 독립적으로 작성하는 것이 가능해 “논리의 세계에서 결정된다”라고 간주한다.</p>
<h3>테이블의 개념</h3>
<p>테이블은 단순히 행과 열만을 가진 것만이 아닌, “공통적인 요소의 집합”임. 관련성이 없는 것이 테이블에 모여서는 안 된다. 또한 테이블은 사람이 만든 개념이나 집합에 대응하는 형식으로 존재해야 한다.</p>
<h2>테이블 설계 규칙</h2>
<h3>사물과 사물의 집합은 계층이 다르다</h3>
<p>테이블 설계의 안티 패턴 중 하나는 “사물과 집합의 계층성”을 무시하는 경우다.</p>
<p>완두콩 테이블</p>

이름 | 색
-- | --
완두콩 | 초록

<p>윗부분에 기본키, 아랫부분에 기본키가 아닌 열을 기재한다. 업계코드에 대한 외래키가 걸려 있으며, 이것은 “업계코드 테이블에 정의되지 않으면 고객기업 테이블에 존재할 수 없음”을 뜻한다</p>
<h3>릴레이션십</h3>
<p>엔티티간의 관련성을 표현하는 것이 “외래키의 존재”이다. 해당 열이 다른 테이블의 기본키를 참조하는 것을 의미하며, 이 관련성을 “릴레이션십” 이라 한다. IE 표기법에서는 아래와 같은 규칙과 기호로 테이블 간 레코드 수의 대응관계를 표시한다</p>
<p><img src="https://s3-us-west-2.amazonaws.com/secure.notion-static.com/7b061664-c857-4383-83ca-d36467ade790/%E1%84%89%E1%85%B3%E1%84%8F%E1%85%B3%E1%84%85%E1%85%B5%E1%86%AB%E1%84%89%E1%85%A3%E1%86%BA_2023-07-16_%E1%84%8B%E1%85%A9%E1%84%92%E1%85%AE_6.56.27.png" alt="스크린샷 2023-07-16 오후 6.56.27.png"></p>
<p>업계 테이블에서 업계코드는 복수의 고객기업 id와 연결될 가능성이 있다. 혹은 고객기업 id와 전혀 연결되지 않을 수도 있다. 따라서 0과 새발을 사용한다.</p>
<p>고객기업 테이블에서 업계코드는 반드시 한 개의 레코드로 대응한다. 따라서 1을 사용한다.</p>

**Comment by somefood** [comment]
# 테이블 설계의 기초

- 관계형 데이터에서 데이터를 관리하고 저장하기 위한 그릇은 `테이블`이기에 이 테이블을 잘 설계하는 것이 매우 중요.
- 관계형 데이터베이스가 표준 데이터베이스 제품이 된 이유는 데이터의 정합성을 높이기 위한 `설계 노하우`가 매우 발달했기 때문임
  - 정규형
  - ER 다이어그램
- 테이블은 2차원 표 형태 처럼 열과 행으로 구성되어있음. 이때 데이터는 공통적인 요소의 집합으로 구성되어져 있어야 한다.

## 테이블 구성의 예

### BAD EXAMPLE

행 사이에 공통적인 요소가 존재하지 않음

|열1|열2|열3|열4|
|--|--|--|--|
|오늘 저녁 반찬|술|배추|파|
|이모씨|박모씨|김모씨|최모씨|
|보너스|플레이스테이션4|제주도|온천|
|남동생|080-0000-1111|OO시 XX구 OO동|1000원 대출|

### GOOD EXAMPLE

공통적인 특징을 가진 데이터의 집합

|원자번호|원소기호|원소명|
|--|--|--|
|1|H|수소|
|2|He|헬륨|
|3|Li|리튬|
|4|Be|베릴륨|

## PK (Primary Key의 존재)

> 테이블에 밑줄 그어진 칼럼이 PK를 의미한다.

- 테이블의 각 행에는 `반드시 기본키를 설정할 것`이라는 설계 원칙이 있음
- 이를 통해 중복 행을 허용하지 않게 함
- 기본키는 중복이 없고 변경이 안되는 컬럼을 지정해야 한다. 변경하면 아래의 문제 소지 발생
  - 변경 후 값의 유일성 보증할 수 없음
  - 과거 데이터와의 결합(매칭)이 어려움
- 기본키는 NULL이 들어갈 수 없다

## 정규형 (Normal Form)

- 테이블은 본인의 기준에 따라 열을 정의하고 테이블을 나눌 수 있는데, 이 상태가 지속되면 설계자 외에 알아볼 수 없는 테이블이 될 수도 있음
- 이런 사태를 줄이기 위해 `테이블은 이렇게 정의해야 한다`는 이론이 구축되었고 이것이 정규형이다.
- 정규형은 제1정규형부터 제5정규형까지 있는데 실무에서는 제3정규형까지만 알아도 충분함

### 제1정규형(1NF)

- 레코드의 단일성의 보증한다.
- 테이블 셀이 복합적인 값을 포함하지 않는다.
  - 복합적인 값은 배열 같은 값이 있음
- BEFORE

> 안석환의 과목은 두 개가 나열되어 있어서 1정규화가 필요하다.

|__사원ID__|이름|과목|
|--|--|--|
|S001|홍석주|DB|
|S002|이건회|자바|
|S003|안석환|(스프링, 자바스크립트)|

- AFTER

정규화를 통해 원자성을 맞춰준다.

|__사원ID__|이름|과목|
|--|--|--|
|S001|홍석주|DB|
|S002|이건회|자바|
|S003|안석환|스프링|
|S003|안석환|자바스크립트|

> 1정규형을 했지만, 갱신 이상이 발생할 수 있음

### 테이블은 함수다

- 왜 복합적인 값을 셀에 넣으면 안 되는가
- 복합적인 값을 허용하면 기본키가 있는 행의 값을 고유하게 특정할 수 없음
- 관계형 데이터베이스의 테이블은 기본키를 `입력 값 X`, 그 외 열 값을 `출력 값 Y`로 간주하는 `입력 값과 출력 값의 대응표`이다.
- 기본키와 다른 열 사이에 성립하는 함수적인 유일성을 관계형 데이터베이스에서는 `함수 종속성(Functional Dependency)`라고 함 {X} -> Y 형태로 표현 함
- 이렇 듯 복합적인 값인 경우에는 함수 종속성이 성립하지 않게 되기에 정규화를 해야함

### 제2정규형(2NF)

- 제1정규형을 만족한 상황에서 진행

- EXAMPLE

- BEFORE

고객기업과 주몬번호가 PK로 되어있는 상황

|__고객기업 ID__|__주문번호__|주문접수일|고객기업명|고객기업 규모|
|--|--|--|--|--|
|CA|001|2023/07/15|A 상사|대규모|
|CA|002|2023/07/16|A 상사|대규모|
|CB|001|2023/07/23|B 건설|중규모|
|CB|002|2023/07/24|B 건설|중규모|
|CC|001|2023/07/25|C 화학|소규모|

- 제1정규형을 충족하고 있으나, `부분함수 종속성`이 발생
- 부분함수 종속성은 `기본키를 구성하는 열의 일부에만 함수 종속이 존재`
- {고객기업 ID} -> {고객 기업명}, {고객기업 ID} -> {고객 규모}를 알 수 있음
- 위와 같은 부분함수 종속이 존재하면 해당 키와 종속하는 열을 다른 테이블로 만들어 외부로 꺼내야 한다.
- 또한, B 건설의 이름을 바꾸고자 하다가 (CB, 002)의 수정을 누락하면 `갱신 이상`이 발생할 수 있음

- AFTER

|__고객기업 ID__|__주문번호__|주문접수일|
|--|--|--|
|CA|001|2023/07/15|
|CA|002|2023/07/16|
|CB|001|2023/07/23|
|CB|002|2023/07/24|
|CC|001|2023/07/25|

|__고객기업 ID__|고객기업명|고객기업 규모|
|--|--|--|
|CA|A 상사|대규모|
|CB|B 건설|중규모|
|CC|C 화학|소규모|

### 제3정규형(3NF)

- 아래 테이블에서 업계만 등록하고 싶은데 그럴려면 회사 하나가 필요한 상황
  - NULL로 넣을 수 있겠지만 바람직하지 않음
- 제2정규형은 만족했으나, `추이함수 종속` 발생. 추이 함수도 테이블을 나누어야 함
  - {고객기업 ID} -> {업계 코드} -> {업계명}

- BEFORE

|__고객기업 ID__|고객기업명|고객기업규모|업계코드|업계명|
|--|--|--|--|--|
|CA|A상사|대규모|D001|석유|
|CB|B건설|중규모|D002|건설|
|CC|C화학|소규모|D003|바이오|

- AFTER

아래처럼 분리하면 따로 따로 원하는 데이터를 넣어도 문제 없다.

|__고객기업 ID__|고객기업명|고객기업규모|업계코드(FK)|
|--|--|--|--|
|CA|A상사|대규모|D001|
|CB|B건설|중규모|D002|
|CC|C화학|소규모|D003|

|업계코드|업계명|
|--|--|
|D001|석유|
|D002|건설|
|D003|바이오|

**Comment by seokhwan-an** [comment]
### 데이터를 관리하는 그릇, 테이블

- 관계형 데이터에서 데이터를 관리하고 저장하기 위한 그릇 ⇒ 전체 데이터가 테이블에 포함되고 저장된다.
- 관계형 데이터베이스에서는 테이블을 사용해 데이터를 적절히 다루는 것이 매우 중요하다.
- 대규모 서비스가 되면 저장해야하는 정보가 다양해지는데 이를 생각나는대로만 넣을 수 있게 설계를 하면 문제가 발생한다.

### 관계형 데이터베이스가 주류가 된 이유

- 데이터베이스에서 데이터의 정합성은 매우 중요하다.
- 관계형 데이터베이스가 정합성을 높이기 위해 `설계 노하우`가 발달했다.

### 테이블 설계는 논리의 세계

- 테이블 설계는 데이터를 저장하는 하드웨어나 서버와 같은 물리층과 관계없이 독립적으로 작성하는 것이 가능해서 `논리의 세계에서 결정된다`라고 간주된다.

### 테이블이란

- 테이블은 공통적인 요소의 집합이다. ⇒ 관련성이 없는 것이 모여서는 안 된다.
- 어떤 공통의 속성을 가진 것의 집합 (테이블 설계의 제1 규칙)
    - 테이블명은 반드시 복수형이나 집합명사로 표현해야 한다.

### 테이블 설계 규칙

- 테이블은 현실 세계의 개념이나 집합을 나타낸 것
- 테이블이란 공통점을 가진 사물의 집합을 나타낸 것
    - 사물과 집합의 계층성을 무시하면 잘못된 테이블 설계가 이루어진다.
    - 토마토, 완두콩, 옥수수를 채소라는 테이블에 묶어두는 것이 아닌 각각이 테이블이 되는 경우가 있다
- 테이블 설계는 인간의 자유로운 사고를 반영하기에 다양한 형태의 테이블 설계가 나올 수 있다.
    - 회원의 테이블이 하나로 구성되는 것이 아닌 등급에 따라 분리, 나이에 따라 분리 등등
    - 위와 같은 실수를 방지하기 위해 `가장 상위의 개념집합으로 테이블을 정하는 규칙` 이 있다.
    - 하나의 회원 테이블로 관리하고 조회할 때 SQL을 잘 활용하면 된다.
- 테이블에는 반드시 기본키를 설정해야한다. ⇒ 한 개 테이블의 내용에는 중복 행을 허용하지 않는다.
    - 두 데이터 행에서 열의 속성이 같은 경우는 괜찮다.
    - 다만 모든 열의 값이 같으면 어느 하나의 데이터가 특정이 되지 않는다.
    - 이 때 기본키를 이용한다.

### 열이란 개체의 속성이다.

- 속성은 성질이나 특징과 같은 의미이다.
- 객체지항 언어에 비유하면 다음과 같다.
    - 테이블은 클래스
    - 각 행은 실체화된 인스턴스
    - 테이블은 메소드가 존재하지 않는다 ⇒ 따라서 `메소드를 뺀 클래스`라고도 한다.

### 기본키(Primary Key)

- 데이터베이스 사용 여부와 관계없이 데이터를 고유하게 식별할 수 있는 기본키를 할당하는 것이 데이터관리의 기본이다.
- 기본키는 중복이 되면 안된다.
- 키본키는 null 값을 가지면 안된다.

### 정규형이란

- 테이블 설계는 자율적으로 설계가 가능하지만 이런 경우 관리하기 어려운 테이블의 구조가 설계될 수 있다.
- 이를 방지하고자 정규형(제대로 된 형태, 데이터의 갱신이 발생한 경우에도 부정합이 발생하기 어려운 테이블의 형태)이 등장했습니다.

**제 1 정규형**

- 테이블 셀에 복합적인 값을 포함하지 않는다.
- 이를 반대하는 테이블 설계는 애초에 불가능하다.
- 스칼라 값 이외의 값을 포함하지 않는 테이블

> 💡스칼라 값이란?
더 이상 나누는 것이 불가능한 값
> 
- 레코드의 단일성을 보증한다라는 데이터의 정합성의 관점에서 나온 규칙이다.

**제 2 정규형**

- 제 2 정규형은 `부분함수 종속성`을 제거하는 규칙이다.

> 💡 부분함수 종속성이란?
기본키를 구성하는 열의 일부에만 함수 종속이 존재하는 것을 의미한다.
> 
- 비정규형 테이블을 실무에서 사용하면 `갱신이상`의 문제점이 발생할 수 있습니다.

> 💡 갱신 이상이란?
데이터를 갱신 시에 데이터의 부정합이 발생하는 상황을 의미한다.
> 

**제 3정규형**

- 기본키에 속하지 않은 모든 속성이 기본키에 이행적 함수 종속이 아닌 경우를 의미한다.
- 책의 예시에서는 현재 `고객기업 Id → 업계코드`, `업계코드 → 업계명`, `고객기업 id → 업계명` 의 관계가 있는데 이는 새로운 업계가 생기려면 항상 고객기업 Id 정보를 가져야 하는 문제점이 발생한다.
    - 따라서 테이블을 각각 분리해야한다. 고객기업 ID를 기본키로 하는 고객기업 테이블과 업계코드를 기본키로 하는 업계 테이블로 분리한다.

### ER 다이어그램

- 갱신 이상을 제거하기 위해 정규화를 진행하면 테이블의 개수가 많아진다.
- 이를 한눈에 관리하기 위해 ER 다이어그램이 개발되었다.
    - Entity는 테이블을 Relationship은 테이블 간의 관계를 의미한다.
    - Entity는 pk를 위의 줄에 기입하고 밑으로는 필요한 필드만 기술한다. 단, 이곳에는 fk는 무조건 정의해두어야 한다. 이를 통해 테이블 간의 관계를 기술할 수 있다.
    - 테이블 간의 관계는 어느 한 테이블이 어느 테이블의 속한다로 시작한다. 각 기호가 존재한다.
        - `-` : 한 개를 의미한다.
        - O : 하나도 없음을 의미한다.
        - ⩚ : 2개 이상을 의미한다.
